### PR TITLE
Add option to choose which fold to use as a final predictor

### PR DESCRIPTION
**Pull request by sami-ka**
Hi @pplonski ,

Thanks for this great package!
I use it pretty often so I wanted to add my contribution to it.

I needed to test the difference between taking the average of models fitted on each fold, and looking at the prediction of only the last fold.
This was especially interesting in my case as it was a time series split, and I wanted my final model to be the one trained on the most recent data.

I added a parameter in the AutoML class called `chosen_fold`, which I ultimately set at `-1` in my case to get the model of the last fold.

It's a bit linked to #475.

Feel free to tell me if I should continue working on this evolution!

P.S. : I think the changes in the requirements_dev.txt are needed because the last click versions are not compatible anymore with the pinned version of black. (see https://stackoverflow.com/questions/71673404/importerror-cannot-import-name-unicodefun-from-click)
Probably upgrading black could also be a good move.

### CODE REVIEW

**Comment by pplonski** [comment]
Hi @drskd!

Thank you for contribution. You are the first person that asked for this feature. If there will be more users that need this, then I will merge it.

**Comment by brainmosaik** [comment]
this sounds nice. but isn´t it the same as, just making a shorter test split ? Can i ask for what u use it? Do u use it only for prediction and not for training?

**Comment by sami-ka** [comment]
@brainmosaik In my case I did a time series split with 4 folds, lets say one for each season over the last 12 months. 
It's reassuring to find model hyperparameters that works well on average for every season, but if I'm currently in summer, I would like to have my predictions based on the most recent model with the weights of the model computed on data from the last months.
For now, the behaviour is to take as final prediction the average prediction over the 4 folds. In my example, that corresponds to actually rely partly on a prediction that was made by a model fitted on data from the winter period, which is not completely satisfying if you have some sort of seasonality.

**Comment by brainmosaik** [comment]
> 

Thats sounds, good. So the workflow for this, should be train with chosen_fold=None,  to find model hyperparameters that works well on average for every season.    And than retrain a new model with found hyperparameters chosen_fold=-1 ,to have a good model for the last fold?  hope i got it right?

**Comment by sami-ka** [comment]
The `chosen_fold` parameter is used only at the prediction stage because, as you already noticed, if I do not want to train on some part of the data I should just discard it.  

Here, as it's already the case, AutoML will train each chosen model and hyperparameters on the 4 splits, then look at the average chosen validation metric over 4 validation sets to rank models on the leaderboard, and finally predicts using the model at the top of the leaderboard. 
Each model in the leaderboard in this example has in fact 4 different models trained each on different datasets. For instance, you could have a decision tree of `max_depth=5` as the winner but if you train it on 4 different datasets, you would have 4 different predictions from each as the decision trees would not necessarily all be the same. `AutoML` returns a prediction as the average prediction of these 4 models. 

The chosen_fold parameter would impact only the prediction step. Each model in the leaderboard would still be trained on  4 different datasets, but the prediction would come from only the model trained on the last split


Classic usage:
```
automl = AutoML(
            validation_strategy={
                "validation_type": "custom"
            },
            chosen_fold=None, # default value
        )
automl.fit(X, y, cv=tscv) # no behaviour change
automl.predict_proba(X) # returns the average predicted proba of 4 models, each trained with winning hyperparameters of the leaderboard on splits defined in CV 
```

Custom usage:
```
automl = AutoML(
            validation_strategy={
                "validation_type": "custom"
            },
            chosen_fold=-1, # only used at prediction time 
        )
automl.fit(X, y, cv=tscv) # no behaviour change
automl.predict_proba(X) # returns the predicted proba of the model trained with winning hyperparameters of the leaderboard on last split only 
```


I could indeed use only the last split if I'm interested in the most recent part of the data, but as the model search is really powerful, having only one split to validate is risky in terms of overfitting. Having additional regularization on hyperparameter selection with more splits help to limit the risk of overfitting.

**Comment by brainmosaik** [comment]
thanks for the explanation , i tryed to take the idea a step further , with something like this  :
elif chosen_fold > 0: 
            # Create weights
            WEIGHTS = np.linspace(0.0, 1.0, len(self.learners))
            ##make the weights Exponential
            WEIGHTS = np.exp(WEIGHTS * chosen_fold)
            ##make ALL_WEIGHTS sum to 1 , So we dont need to divide by  float(len(self.learners)) ?  I think this works only , for regression ?
            WEIGHTS = WEIGHTS /  np.sum(WEIGHTS)

            ## Make last weight 1 | so we still need to divide by float(len(self.learners)) . but i think this will take the prediction off.
            #WEIGHTS = WEIGHTS /  WEIGHTS[-1]

            print("CHOSEN FOLD: WEIGHTED",WEIGHTS )
            for ind, learner in enumerate(self.learners):
                # preprocessing goes here
                X_data, _, _ = self.preprocessings[ind].transform(X.copy(), None)
                y_p = learner.predict(X_data)
                y_p = self.preprocessings[ind].inverse_scale_target(y_p)

                ##multiply Prediction with WEIGHT
                y_p = y_p * WEIGHTS[ind] 
                
                y_predicted = y_p if y_predicted is None else y_predicted + y_p

            y_predicted_average = y_predicted #/ float(len(self.learners))

            y_predicted_final = self.preprocessings[0].prepare_target_labels(
                y_predicted_average
            )

**Comment by mosaikme** [comment]
i must say thanks again, for this commit. I wonder if this got , implemented now in the main branch?
